src/ripple/shamap/SHAMapInnerNode.h

```diff
@@ -38,15 +38,194 @@ namespace ripple {
 class SHAMapInnerNode final : public SHAMapAbstractNode,
                               public CountedObject<SHAMapInnerNode>
 {
-    std::array<SHAMapHash, 16> mHashes;
-    std::shared_ptr<SHAMapAbstractNode> mChildren[16];
-    int mIsBranch = 0;
-    std::uint32_t mFullBelowGen = 0;
+public:
+    /** Each inner node has 16 children (the 'radix tree' part of the map) */
+    static inline constexpr unsigned int branchFactor = 16;
+
+    // TaggedPointer is a combination of a pointer and a mask stored in the
+    // lowest two bits. Since pointers do not have arbitrary alignment, the
+    // lowest bits in the pointer are guaranteed to be zero. TaggedPointer
+    // stores information in these low bits. When dereferencing the pointer,
+    // these low "tag" bits are set to zero. When accessing the tag bits, the
+    // high "pointer" bits are set to zero.
+    //
+    // The "pointer" part points to to the equivalent to an array of
+    // `SHAMapHash` followed immediately by an array of
+    // `shared_ptr<SHAMapAbstractNode>`. The sizes of these arrays are
+    // determined by the tag. The tag is an index into an array (`boundaries`,
+    // defined in the cpp file) that specifies the size. Both arrays are the
+    // same size. Note that the sizes may be smaller than the full 16 elements
+    // needed to explicitly store all the children. In this case, the arrays
+    // only store the non-empty children. The non-empty children are stored in
+    // index order. For example, if only children `2` and `14` are non-empty, a
+    // two-element array would store child `2` in array index 0 and child `14`
+    // in array index 1. There are functions to convert between a child's tree
+    // index and the child's index in a sparse array.
+    //
+    // The motivation for this class is saving RAM. A large percentage of inner
+    // nodes only store a small number of children. Memory can be saved by
+    // storing the inner node's children in sparse arrays. Measurements show
+    // that on average a typical SHAMap's inner nodes can be stored using only
+    // 25% of the original space.
+
+    class TaggedPointer
```

Comment:
Agreed about the doxygen-style comments, Fixed in `2fa4a8d59e [fold] Doxygen comments and misc cleanups 1:`

Let me think about moving `TaggedPointer` into the cpp file. That class started out as a set of free functions inside the cpp file, but I decided to package together `uintptr_t` and the related functions into that `TaggedPointer` class. If we do decide to move that back into the cpp, I'd probably use free functions again. I'm don't love creating a `TaggedPointer` object just to call member functions like that.

I think I prefer the inner-class implementation, but since you brought it up, let me think about it some more.

Reply:
I'm fine either way; I just think that having a `class` in the **.cpp** abstracts away the details of the class, since its interface is never intended to be public and only pollutes here.

Having free functions works, but, at the same time, I think having the class emphasizes that there's some kind of structured type being operated on.

At your discretion, really.

Reply:
@nbougalis I decided to move the `TaggedPointer` class and split it into a header file included into SHAMapInnerNode.h and a `.ipp` implementation file included into SHAMapInnerNode.cpp. (see the patch `713b7f7612 [FOLD] Move TaggedPoint to its own file`.

I didn't like having the class in the cpp because we'd have to create the class from the opaque `uintptr`, which makes the code a little awkward. I also prefer the class to free functions. I think this .h/.ipp split is a reasonable solution.